File: sdk/node/src/updater_facade.ts
/** Generated host facade (relkit.updater.v1). Do not hand-edit the method set. */

import { spawn } from "node:child_process";
import { Buffer } from "node:buffer";
import { existsSync } from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

import { create, fromBinary, toBinary } from "@bufbuild/protobuf";
import {
  ApplyOpSchema,
  ApplyResultSchema,
  CheckOpSchema,
  CheckResultSchema,
  CleanupOpSchema,
  DownloadOpSchema,
  DownloadResultSchema,
  ErrorSchema,
  FailedSchema,
  ListInstalledOpSchema,
  ResultSchema,
  RollbackOpSchema,
  SkipOpSchema,
  StatusOpSchema,
  SwitchActiveOpSchema,
  UpdaterEventSchema,
  UpdaterRequestSchema,
  type Capabilities,
  type CheckPolicy,
  type CheckResult,
  type ClientProfile,
  type DownloadResult,
  type ApplyResult,
  type Error,
  type Runtime,
  type StatusSnapshot,
  type UpdaterEvent,
  type UpdaterRequest,
  type Result,
} from "./gen/updater/v1/updater_pb.js";

export const ipcMin = 1;
export const ipcMax = 2;
export const ipcCurrent = 2;

const facadeError = (code: number, message: string, retryable = false): Error =>
  create(ErrorSchema, { code, retryable, message, attempts: [] });

export type Glue = {
  locate(runtime: Runtime): Promise<string>;
  run(bin: string, args: string[], stdin: Uint8Array): AsyncIterable<UpdaterEvent>;
};

function frame(bytes: Uint8Array): Uint8Array {
  const out = Buffer.alloc(4 + bytes.length);
  out.writeUInt32BE(bytes.length, 0);
  Buffer.from(bytes).copy(out, 4);
  return out;
}

async function* decodeFrames(stream: AsyncIterable<Buffer>): AsyncGenerator<UpdaterEvent> {
  let buf = Buffer.alloc(0);
  for await (const chunk of stream) {
    buf = Buffer.concat([buf, chunk]);
    while (buf.length >= 4) {
      const n = buf.readUInt32BE(0);
      if (buf.length < 4 + n) break;
      const payload = buf.subarray(4, 4 + n);
      buf = buf.subarray(4 + n);
      yield fromBinary(UpdaterEventSchema, payload);
    }
  }
}

export const defaultGlue: Glue = {
  async locate(runtime) {
    const name = process.platform === "win32" ? "relkit-updater.exe" : "relkit-updater";
    const candidates: string[] = [];
    if (runtime.sidecarPath) candidates.push(runtime.sidecarPath);
    if (process.env.RELKIT_UPDATER) candidates.push(process.env.RELKIT_UPDATER);
    if (runtime.install?.installRoot) {
      candidates.push(path.join(runtime.install.installRoot, name));
    }
    candidates.push(path.join(path.dirname(process.execPath), name));
    for (const c of candidates) {
      if (c && existsSync(c)) return c;
    }
    throw new Error("sidecar not found");
  },
  async *run(bin, args, stdin) {
    const child = spawn(bin, args, { stdio: ["pipe", "pipe", "inherit"] });
    child.stdin.end(Buffer.from(stdin));
    yield* decodeFrames(child.stdout);
    await new Promise<void>((resolve, reject) => {
      child.on("error", reject);
      child.on("close", () => resolve());
    });
  },
};

export class Updater {
  constructor(
    private readonly profile: ClientProfile,
    private readonly runtime: Runtime,
    private readonly glue: Glue,
    private readonly bin: string,
    readonly capabilities: Capabilities,
  ) {}

  static async open(
    profile: ClientProfile,
    runtime: Runtime,
    glue: Glue = defaultGlue,
  ): Promise<{ kind: "opened"; updater: Updater; capabilities: Capabilities } | { kind: "failed"; error: Error }> {
    try {
      const bin = await glue.locate(runtime);
      let caps: Capabilities | undefined;
      let ferr: Error | undefined;
      for await (const ev of glue.run(bin, ["-capabilities"], new Uint8Array())) {
        if (ev.kind.case === "capabilities") caps = ev.kind.value;
        if (ev.kind.case === "failed") ferr = ev.kind.value.error;
      }
      if (ferr) return { kind: "failed", error: ferr };
      if (!caps) {
        return {
          kind: "failed",
          error: facadeError(8, "no capabilities"),
        };
      }
      if (caps.ipc < ipcMin) {
        return { kind: "failed", error: facadeError(9, "sidecar IPC too old") };
      }
      if (caps.ipc > ipcMax) {
        return { kind: "failed", error: facadeError(10, "sidecar IPC too new") };
      }
      return { kind: "opened", updater: new Updater(profile, runtime, glue, bin, caps), capabilities: caps };
    } catch (e) {
      return {
        kind: "failed",
        error: facadeError(19, String(e)),
      };
    }
  }

  private async *call(partial: Partial<UpdaterRequest> & { op?: UpdaterRequest["op"] }): AsyncGenerator<UpdaterEvent> {
    const req = {
      hello: { ipcMin, ipcMax },
      profile: this.profile,
      runtime: this.runtime,
      ...partial,
    } as UpdaterRequest;
    const bytes = frame(toBinary(UpdaterRequestSchema, req));
    yield* this.glue.run(this.bin, [], bytes);
  }

  async check(opts: { force?: boolean; exactCode?: bigint; policy?: CheckPolicy } = {}): Promise<CheckResult> {
    for await (const ev of this.call({
      op: { case: "check", value: create(CheckOpSchema, { force: !!opts.force, exactCode: opts.exactCode ?? 0n, policy: opts.policy }) },
    })) {
      if (ev.kind.case === "check") return ev.kind.value;
      if (ev.kind.case === "failed") return { kind: { case: "failed", value: ev.kind.value } } as CheckResult;
    }
    return create(CheckResultSchema, {
      kind: { case: "failed", value: create(FailedSchema, {
        error: facadeError(1, "no check result", true),
      }) },
    });
  }

  async skip(opts: { code: bigint }): Promise<Result> {
    for await (const ev of this.call({ op: { case: "skip", value: create(SkipOpSchema, { code: opts.code }) } })) {
      if (ev.kind.case === "result") return ev.kind.value;
    }
    return create(ResultSchema, { kind: { case: "ok", value: {} } });
  }

  async download(opts: { planId: string }, onEvent?: (e: UpdaterEvent) => void): Promise<DownloadResult> {
    for await (const ev of this.call({ op: { case: "download", value: create(DownloadOpSchema, { planId: opts.planId }) } })) {
      onEvent?.(ev);
      if (ev.kind.case === "download") return ev.kind.value;
    }
    return create(DownloadResultSchema, {
      kind: { case: "failed", value: create(FailedSchema, {
        error: facadeError(1, "no download result", true),
      }) },
    });
  }

  async apply(opts: { planId: string; installOnly?: boolean }, onEvent?: (e: UpdaterEvent) => void): Promise<ApplyResult> {
    for await (const ev of this.call({ op: { case: "apply", value: create(ApplyOpSchema, { planId: opts.planId, installOnly: !!opts.installOnly }) } })) {
      onEvent?.(ev);
      if (ev.kind.case === "apply") return ev.kind.value;
    }
    return create(ApplyResultSchema, {
      kind: { case: "failed", value: create(FailedSchema, {
        error: facadeError(1, "no apply result", true),
      }) },
    });
  }

  async status(): Promise<StatusSnapshot> {
    for await (const ev of this.call({ op: { case: "status", value: create(StatusOpSchema) } })) {
      if (ev.kind.case === "status") return ev.kind.value;
    }
    return {} as StatusSnapshot;
  }

  async cleanup(): Promise<Result> {
    for await (const ev of this.call({ op: { case: "cleanup", value: create(CleanupOpSchema) } })) {
      if (ev.kind.case === "result") return ev.kind.value;
    }
    return create(ResultSchema, { kind: { case: "ok", value: {} } });
  }

  async cancel(): Promise<Result> {
    return create(ResultSchema, { kind: { case: "ok", value: {} } });
  }

  async listInstalled(): Promise<import("./gen/updater/v1/updater_pb.js").InstalledList> {
    for await (const ev of this.call({ op: { case: "listInstalled", value: create(ListInstalledOpSchema) } })) {
      if (ev.kind.case === "installed") return ev.kind.value;
    }
    return { versions: [], activeCode: 0n } as import("./gen/updater/v1/updater_pb.js").InstalledList;
  }

  async switchActive(opts: { code: bigint }): Promise<Result> {
    for await (const ev of this.call({ op: { case: "switchActive", value: create(SwitchActiveOpSchema, { code: opts.code }) } })) {
      if (ev.kind.case === "result") return ev.kind.value;
    }
    return create(ResultSchema, { kind: { case: "ok", value: {} } });
  }

  async rollback(): Promise<Result> {
    for await (const ev of this.call({ op: { case: "rollback", value: create(RollbackOpSchema) } })) {
      if (ev.kind.case === "result") return ev.kind.value;
    }
    return create(ResultSchema, { kind: { case: "ok", value: {} } });
  }
}

void fileURLToPath;

File: sdk/dart/lib/src/updater_facade.dart
/// Generated host facade (relkit.updater.v1). Do not hand-edit the method set.
library;

import 'dart:core' hide Error;
import 'dart:io';
import 'dart:typed_data';

import 'package:fixnum/fixnum.dart';
import 'package:protobuf/protobuf.dart';

import 'gen/updater/v1/updater.pb.dart';

const int ipcMin = 1;
const int ipcMax = 1;
const int ipcCurrent = 1;

abstract class Glue {
  Future<String> locate(Runtime runtime);
  Stream<UpdaterEvent> run(
    String bin,
    List<String> args,
    Uint8List stdin,
  );
}

class DefaultGlue implements Glue {
  @override
  Future<String> locate(Runtime runtime) async {
    final name =
        Platform.isWindows ? 'relkit-updater.exe' : 'relkit-updater';
    final candidates = <String>[];
    if (runtime.sidecarPath.isNotEmpty) {
      candidates.add(runtime.sidecarPath);
    }
    final env = Platform.environment['RELKIT_UPDATER'];
    if (env != null && env.isNotEmpty) candidates.add(env);
    if (runtime.hasInstall() && runtime.install.installRoot.isNotEmpty) {
      candidates.add('${runtime.install.installRoot}${Platform.pathSeparator}$name');
    }
    candidates.add('${File(Platform.resolvedExecutable).parent.path}${Platform.pathSeparator}$name');
    for (final c in candidates) {
      if (File(c).existsSync()) return c;
    }
    throw StateError('sidecar not found');
  }

  @override
  Stream<UpdaterEvent> run(
    String bin,
    List<String> args,
    Uint8List stdinBytes,
  ) async* {
    final proc = await Process.start(bin, args);
    proc.stdin.add(stdinBytes);
    await proc.stdin.close();
    proc.stderr.listen((_) {});
    yield* _readFrames(proc.stdout);
    await proc.exitCode;
  }
}

Stream<UpdaterEvent> _readFrames(Stream<List<int>> stdout) async* {
  final buf = BytesBuilder(copy: false);
  await for (final chunk in stdout) {
    buf.add(chunk);
    while (true) {
      final data = buf.toBytes();
      if (data.length < 4) break;
      final n = ByteData.sublistView(data).getUint32(0, Endian.big);
      if (data.length < 4 + n) break;
      final payload = data.sublist(4, 4 + n);
      final rest = data.sublist(4 + n);
      buf.clear();
      buf.add(rest);
      yield UpdaterEvent.fromBuffer(payload);
    }
  }
}

Uint8List _frame(GeneratedMessage msg) {
  final payload = msg.writeToBuffer();
  final out = BytesBuilder(copy: false);
  final hdr = ByteData(4)..setUint32(0, payload.length, Endian.big);
  out.add(hdr.buffer.asUint8List());
  out.add(payload);
  return out.takeBytes();
}

class OpenResult {
  OpenResult.opened(this.updater, this.capabilities)
      : error = null,
        kind = 'opened';
  OpenResult.failed(this.error)
      : updater = null,
        capabilities = null,
        kind = 'failed';

  final String kind;
  final Updater? updater;
  final Capabilities? capabilities;
  final Error? error;
}

class Updater {
  Updater._(this._profile, this._runtime, this._glue, this._bin, this.caps);

  final ClientProfile _profile;
  final Runtime _runtime;
  final Glue _glue;
  final String _bin;
  final Capabilities caps;

  static Future<OpenResult> open(
    ClientProfile profile,
    Runtime runtime, {
    Glue? glue,
  }) async {
    glue ??= DefaultGlue();
    try {
      final bin = await glue.locate(runtime);
      Capabilities? caps;
      Error? err;
      await for (final ev in glue.run(bin, ['-capabilities'], Uint8List(0))) {
        if (ev.hasCapabilities()) caps = ev.capabilities;
        if (ev.hasFailed()) err = ev.failed.error;
      }
      if (err != null) {
        return OpenResult.failed(err);
      }
      if (caps == null) {
        return OpenResult.failed(
            Error(code: ErrorCode.ERROR_CODE_PROTOCOL_MISMATCH, message: 'no capabilities'));
      }
      if (caps.ipc < ipcMin) {
        return OpenResult.failed(
            Error(code: ErrorCode.ERROR_CODE_UPDATER_TOO_OLD, message: 'sidecar IPC too old'));
      }
      if (caps.ipc > ipcMax) {
        return OpenResult.failed(
            Error(code: ErrorCode.ERROR_CODE_UPDATER_TOO_NEW, message: 'sidecar IPC too new'));
      }
      return OpenResult.opened(
          Updater._(profile, runtime, glue, bin, caps), caps);
    } on Object catch (e) {
      return OpenResult.failed(Error(
          code: ErrorCode.ERROR_CODE_SIDECAR_NOT_FOUND, message: e.toString()));
    }
  }

  Stream<UpdaterEvent> _call(UpdaterRequest req) {
    req.hello = ClientHello(ipcMin: ipcMin, ipcMax: ipcMax);
    req.profile = _profile;
    req.runtime = _runtime;
    return _glue.run(_bin, const <String>[], _frame(req));
  }

  /// protobuf.dart reserves `check` on GeneratedMessage, so the field is `check_10`.
  Future<CheckResult> check({bool force = false, int exactCode = 0, CheckPolicy? policy}) async {
    final req = UpdaterRequest()
      ..check_10 = CheckOp(force: force, exactCode: Int64(exactCode), policy: policy);
    await for (final ev in _call(req)) {
      if (ev.hasCheck_10()) return ev.check_10;
      if (ev.hasFailed()) {
        return CheckResult()..failed = ev.failed;
      }
    }
    return CheckResult()
      ..failed = Failed(
          error: Error(
              code: ErrorCode.ERROR_CODE_NETWORK, message: 'no check result'));
  }

  Future<Result> skip({required int code}) async {
    final req = UpdaterRequest()..skip = SkipOp(code: Int64(code));
    await for (final ev in _call(req)) {
      if (ev.hasResult()) return ev.result;
    }
    return Result()..ok = Ok();
  }

  Future<DownloadResult> download({
    required String planId,
    void Function(UpdaterEvent event)? onEvent,
  }) async {
    final req = UpdaterRequest()..download = DownloadOp(planId: planId);
    await for (final ev in _call(req)) {
      onEvent?.call(ev);
      if (ev.hasDownload()) return ev.download;
    }
    return DownloadResult()
      ..failed = Failed(
          error: Error(
              code: ErrorCode.ERROR_CODE_NETWORK, message: 'no download result'));
  }

  Future<ApplyResult> apply({
    required String planId,
    void Function(UpdaterEvent event)? onEvent,
  }) async {
    final req = UpdaterRequest()..apply = ApplyOp(planId: planId);
    await for (final ev in _call(req)) {
      onEvent?.call(ev);
      if (ev.hasApply()) return ev.apply;
    }
    return ApplyResult()
      ..failed = Failed(
          error: Error(
              code: ErrorCode.ERROR_CODE_NETWORK, message: 'no apply result'));
  }

  Future<StatusSnapshot> status() async {
    final req = UpdaterRequest()..status = StatusOp();
    await for (final ev in _call(req)) {
      if (ev.hasStatus()) return ev.status;
    }
    return StatusSnapshot();
  }

  Future<Result> cleanup() async {
    final req = UpdaterRequest()..cleanup = CleanupOp();
    await for (final ev in _call(req)) {
      if (ev.hasResult()) return ev.result;
    }
    return Result()..ok = Ok();
  }

  Future<Result> cancel() async {
    return Result()..ok = Ok();
  }
}
